### PR TITLE
Remove optional SHA-1 in the default TLS configuration.

### LINKED ISSUE #4171: Remove MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES
The compile-time option `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES` was intended as a temporary thing to facilitate the transition in environments with SHA-1 certificates. It's been over 3 years, and it's way past time it got removed. Everyone should have upgraded their infrastructure to support and use SHA2 by now.

As this is a security concern, I'm bypassing mailing list discussion. `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES` must go. When we introduced it, we only intended to keep it for a few months, we didn't think it would stay until the next major version.

Part of what slowed us down is that it took us a long time to migrate our test data, and I'm not sure we've done everything. So the goals for this task are:

* Remove `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES`, the one line of code that it guards, and all test cases that have it as a dependency. All of these test cases are specifically about the effect of the option so there's no loss of coverage here.
* Document the migration as using a custom X.509 verification profile if you really need to use a SHA-1 certificate.
* It's possible that we have some testing that depends dynamically on which algorithms are permitted, and since SHA-1 is no longer allowed in the `full` configuration, these tests no longer run. Review the tests to see if there are any such cases. At least compare `outcomes.csv` on a CI run before and after the removal of `MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES`.

### PR DESCRIPTION
**Pull request by mstarzyk-mobica**
Resolved #4171

### CODE REVIEW

File: ChangeLog.d/remove_allow_sha1_in_certificates [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Removals
+   * Remove optional SHA-1 in the default TLS configuration for certificate
+     signing. This feature was ment to be available only temporarily.
+     Users are expected to use SHA-2 instead, since SHA-1 is currently
+     considered a security risk.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/remove_allow_sha1_in_certificates:5]: The first sentence is hard to understand. What we're doing here is removing the compile-time option that changes the default behavior of TLS connections to allow SHA-1-signed certificates. It is not removing the permission to use SHA-1 by default: that was off by default a long time ago. Suggested rewording:

> Remove the compile-time option MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES, which was off by default. Users should not trust certificates signed with SHA-1 due to the known attacks against SHA-1.

(I think the name of the option is clear enough not to require additional explanation.)

File: ChangeLog.d/remove_allow_sha1_in_certificates [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+     If needed, SHA-1 cerificate can still be used by providing custom
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/remove_allow_sha1_in_certificates:6]: I think this level of detail belongs in a separate place (a knowledge base entry): it's too much information for the changelog. But I'm ok with having it here now and moving it later (before the 3.0 release!).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM except for the changelog entry

File: ChangeLog.d/remove_allow_sha1_in_certificates (hunk 2 above)
Comment (mstarzyk-mobica) [reply ChangeLog.d/remove_allow_sha1_in_certificates:5]: @daverodgman suggested this is a bit too niche to include in mbedTLS docs.

**Review by gilles-peskine-arm — DISMISSED** [review]

**Review by TRodziewicz — APPROVED** [review]
Looks good to me.

Although a conflict needs to be resolved:
Conflicting files
tests/scripts/all.sh

**Comment by mstarzyk-mobica** [comment]
fixed conflict with rebase

File: tests/scripts/all.sh [hunk 3]
```diff
@@ -2499,13 +2499,10 @@ component_build_armcc () {
...
 }
 
-component_test_allow_sha1 () {
-    msg "build: allow SHA1 in certificates by default"
-    scripts/config.py set MBEDTLS_TLS_DEFAULT_ALLOW_SHA1_IN_CERTIFICATES
-    make CFLAGS='-Werror -Wall -Wextra'
-    msg "test: allow SHA1 in certificates by default"
-    make test
-    if_build_succeeded tests/ssl-opt.sh -f SHA-1
+component_build_ssl_hw_record_accel() {
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:2502]: Rebase mistake: `component_build_ssl_hw_record_accel` is being re-added.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]

File: tests/scripts/all.sh (hunk 3 above)
Comment (mstarzyk-mobica) [reply tests/scripts/all.sh]: rebased again. thanks for spotting this

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by TRodziewicz — APPROVED** [review]

**Pull request merged**